### PR TITLE
Merge pull request #78307 from blanky0230/nixos-tuxedo-keyboard

### PR DESCRIPTION
**Pull request by blanky0230**
###### Motivation for this change

Additional hardware-support for another hardware vendor. 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers
@Infinisil this is a replication of  #73885.  I completely messed up when trying to remove the merge commit :roll_eyes: I'm sorry to bother you... Feel free to close this, if I'd been to noisy :|

### CODE REVIEW

File: pkgs/os-specific/linux/tuxedo-keyboard/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  src = fetchFromGitHub {
+    owner = "tuxedocomputers";
+    repo = "tuxedo-keyboard";
+    rev = "d65e76e84cfd8169591fc2a0a7c9219fa19da1b5";
+    sha256 = "1s48qpwybwh5pwqas2d1v2a7x4r97sm4hr9i4902r1d7h384bv17";
+  };
+
+  preConfigure = ''
+    substituteInPlace Makefile --replace '/lib/modules/' "${kernel.dev}/lib/modules/"
+    substituteInPlace Makefile --replace '$(shell uname -r)' "${kernel.modDirVersion}"
```
Comment (balsoft) [inline pkgs/os-specific/linux/tuxedo-keyboard/default.nix]: Can't this be replaced with `makeFlags = [ "KDIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build" ];` ?

**Review by balsoft — COMMENTED** [review]
Hi! Thank you very for contributing. It's always nice to have more hardware support.

You have a typo in commit message

> tuxedo-keyboar: create and register module for tuxedo-keyboard driver.

**Comment by balsoft** [comment]
@GrahamcOfBorg build tuxedo-keyboard

File: pkgs/os-specific/linux/tuxedo-keyboard/default.nix (hunk 1 above)
Comment (blanky0230) [reply pkgs/os-specific/linux/tuxedo-keyboard/default.nix]: That works like a charm! Thanks!

**Comment by balsoft** [comment]
@GrahamcOfBorg build tuxedo-keyboard

@blanky0230 I don't have push access if that's what you're asking for.

**Comment by blanky0230** [comment]
@balsoft Not at all. I didn't mean to bug you. I just thought it'd be sensible to ask again what you thought.

File: nixos/modules/hardware/tuxedo-keyboard.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+with lib;
+
+let 
+  cfg = config.hardware.tuxedo-keyboard;
+  tuxedo-keyboard = config.boot.kernelPackages.tuxedo-keyboard;
+in
+  {
+    options.hardware.tuxedo-keyboard = {
+      enable = mkOption {
```
Comment (balsoft) [inline nixos/modules/hardware/tuxedo-keyboard.nix]:
```suggestion
      enable = mkEnableOption ''
        the tuxedo-keyboard driver.
        To configure the driver, pass the options to the <option>boot.kernelParams</option> configuration.
        There are several parameters you can change. It's best to check at the source code description which options are supported.
        You can find all the supported parameters at: <link xlink:href="https://github.com/tuxedocomputers/tuxedo-keyboard#kernelparam" />
         In order to use the <literal>custom</literal> lighting with the maximumg brightness and a color of <literal>0xff0a0a</literal> one would put pass <option>boot.kernelParams</option> like this:
        <programlisting>
        boot.kernelParams = [
         "tuxedo_keyboard.mode=0"
         "tuxedo_keyboard.brightness=255"
         "tuxedo_keyboard.color_left=0xff0a0a"
        ];
        </programlisting>
      '';
```

Comment (balsoft) [inline nixos/modules/hardware/tuxedo-keyboard.nix]:
```suggestion
      enable = mkEnableOption ''
          Enables the tuxedo-keyboard driver.
          
          To configure the driver, pass the options to the <option>boot.kernelParams</option> configuration.
          There are several parameters you can change. It's best to check at the source code description which options are supported.
          You can find all the supported parameters at: <link xlink:href="https://github.com/tuxedocomputers/tuxedo-keyboard#kernelparam" />
          In order to use the <literal>custom</literal> lighting with the maximumg brightness and a color of <literal>0xff0a0a</literal> one would put pass <option>boot.kernelParams</option> like this:
          <programlisting>
          boot.kernelParams = [
           "tuxedo_keyboard.mode=0"
           "tuxedo_keyboard.brightness=255"
           "tuxedo_keyboard.color_left=0xff0a0a"
          ];
          </programlisting>
        '';
```

(Fix the formatting yourself, I'm too lazy to do that in awful github interface)

File: pkgs/os-specific/linux/tuxedo-keyboard/default.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{ stdenv, fetchFromGitHub, kernel, kmod }:
+
+stdenv.mkDerivation rec {
+  name = "tuxedo-keyboard-${kernel.version}-${version}";
```
Comment (balsoft) [inline pkgs/os-specific/linux/tuxedo-keyboard/default.nix]:
```suggestion
  pname = "tuxedo-keyboard-${kernel.version}";
```

**Review by balsoft — APPROVED** [review]
Otherwise, looks good.

**Comment by balsoft** [comment]
@GrahamcOfBorg build linuxPackages.tuxedo-keyboard

**Comment by balsoft** [comment]
Commit names should be 
```
tuxedo-keyboard: init at 2019-08-26
nixos/tuxedo-keyboard: init module 
```

**Review by infinisil — APPROVED** [review]
Nice! Looking good!

**Pull request merged**